applications/StructuralMechanicsApplication/custom_elements/timoshenko_beam_element_2D2N.cpp

```diff
@@ -0,0 +1,1018 @@
+    auto &r_cl_options = cl_values.GetOptions();
+    r_cl_options.Set(ConstitutiveLaw::COMPUTE_STRESS             , true);
+    r_cl_options.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, true);
+
+    const double Phi    = CalculatePhi(cl_values);
+    const double length = CalculateLength();
+    const double J      = 0.5 * length;
+    const double area   = GetProperties()[CROSS_AREA];
+
+    // Let's initialize the cl values
+    VectorType strain_vector(3), stress_vector(3);
+    MatrixType constitutive_matrix(3, 3);
+    strain_vector.clear();
+    cl_values.SetStrainVector(strain_vector);
+    cl_values.SetStressVector(stress_vector);
+    cl_values.SetConstitutiveMatrix(constitutive_matrix);
+    VectorType nodal_values(mat_size);
+    GetNodalValuesVector(nodal_values);
+    VectorType global_size_N(mat_size), N_u_derivatives(mat_size-4), N_theta_derivatives(mat_size-2), N_theta(mat_size-2), N_derivatives(mat_size-2), N_u(mat_size-4), N_shape(mat_size-2), N_s(mat_size-2);
+
+    // Loop over the integration points
+    for (SizeType IP = 0; IP < integration_points.size(); ++IP) {
+        const auto local_body_forces = GetLocalAxesBodyForce(*this, integration_points, IP);
+
+        global_size_N.clear();
+        const double xi     = integration_points[IP].X();
+        const double weight = integration_points[IP].Weight();
+        const double jacobian_weight = weight * J;
+
+        CalculateGeneralizedStrainsVector(strain_vector, length, Phi, xi, nodal_values);
+
+        mConstitutiveLawVector[IP]->CalculateMaterialResponseCauchy(cl_values);
+        const Vector &r_generalized_stresses = cl_values.GetStressVector();
+        const double N = r_generalized_stresses[0];
+        const double M = r_generalized_stresses[1];
+        const double V = r_generalized_stresses[2];
+
+        const MatrixType& r_constitutive_matrix = cl_values.GetConstitutiveMatrix();
+        const double dN_dEl    = r_constitutive_matrix(0, 0);
+        const double dM_dkappa = r_constitutive_matrix(1, 1);
+        const double dV_dgamma = r_constitutive_matrix(2, 2);
+
+        GetFirstDerivativesNu0ShapeFunctionsValues(N_u_derivatives, length, Phi, xi);
+        GetFirstDerivativesNThetaShapeFunctionsValues(N_theta_derivatives, length, Phi, xi);
+        GetNThetaShapeFunctionsValues(N_theta, length, Phi, xi);
+        GetFirstDerivativesShapeFunctionsValues(N_derivatives, length, Phi, xi);
+        GetShapeFunctionsValues(N_shape, length, Phi, xi);
+        GetNu0ShapeFunctionsValues(N_u, length, Phi, xi);
+        noalias(N_s) = N_derivatives - N_theta;
+
+        // Axial contributions
+        GlobalSizeAxialVector(global_size_N, N_u_derivatives);
+        noalias(rLHS) += outer_prod(global_size_N, global_size_N) * dN_dEl * jacobian_weight;
+        noalias(rRHS) -= global_size_N * N * jacobian_weight;
+
+        if (r_constitutive_matrix(0, 1) != 0.0 || r_constitutive_matrix(0, 2) != 0.0) {
```

Comment:
this if is slower than doing the multiplicatino. Honestly i would just multiply ...

Reply:
done